controllers/config.go

```diff
@@ -0,0 +1,231 @@
+	"fmt"
+
+	"github.com/ghodss/yaml"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	odfOperatorConfigmapName = "odf-operator-manager-config"
+)
+
+type ConfigData struct {
+	/* example
+	   channel: alpha
+	   csv: ocs-operator.v4.18.0
+	   pkg: ocs-operator
+	   scalerCrds:
+	     - storageclusters.ocs.openshift.io
+	   ---------------------------------------
+	   channel: beta
+	   csv: odf-prometheus-operator.v4.18.0
+	   pkg: odf-prometheus-operator
+	   scalerCrds:
+	     - alertmanagers.monitoring.coreos.com
+	     - prometheuses.monitoring.coreos.com
+	*/
+
+	Channel    string   `yaml:"channel"`
+	Csv        string   `yaml:"csv"`
+	Pkg        string   `yaml:"pkg"`
+	ScalerCrds []string `yaml:"scalerCrds"`
+}
+
+func GetConfig(ctx context.Context, cli client.Client, logger logr.Logger, operatorNamespace string) (map[string]ConfigData, error) {
+
+	// read configmap
+	cm := &corev1.ConfigMap{}
+	cm.Name = odfOperatorConfigmapName
+	cm.Namespace = operatorNamespace
+	if err := cli.Get(ctx, client.ObjectKeyFromObject(cm), cm); err != nil {
+		logger.Error(err, "failed to get configmap")
+		return nil, err
+	}
+
+	// parse the ConfigMap data and skip any keys that fail to parse
+	configMapData := make(map[string]ConfigData)
+	for key, value := range cm.Data {
+
+		// skip parsing known environment variable keys from the configmap.
+		// first condition can be removed once the older keys are removed from the configmap.
+		if _, ok := DefaultValMap[key]; ok || key == "controller_manager_config.yaml" {
+			continue
+		}
+
+		var config ConfigData
+		if err := yaml.Unmarshal([]byte(value), &config); err != nil {
+			logger.Error(err, "failed to unmarshal configmap data", "key", key)
+			continue
+		}
+		configMapData[key] = config
+	}
+
+	logger.Info("parsed configmap data successfully", "configMapData", configMapData)
+
+	return configMapData, nil
+}
+
+type CrdAndPackages struct {
+	/* examples
+	   CrdName:    "storageclusters.ocs.openshift.io",
+	   ApiVersion: "ocs.openshift.io/v1",
+	   Kind:       "StorageCluster",
+	   PkgNames:   []string{OcsSubscriptionPackage},
+	   -----------------------------------------------
+	   CrdName:    "cephclusters.ceph.rook.io",
+	   ApiVersion: "ceph.rook.io/v1",
+	   Kind:       "CephCluster",
+	   PkgNames:   []string{RookSubscriptionPackage, CephCSISubscriptionPackage, CSIAddonsSubscriptionPackage},
+	*/
+
+	CrdName    string
+	ApiVersion string
+	Kind       string
+	CrdPresent bool
+	PkgNames   []string
+}
+
+func GetCrdAndPackagesMappingList(ctx context.Context, cli client.Client, logger logr.Logger, operatorNamespace string) ([]CrdAndPackages, error) {
+
+	configMapData, err := GetConfig(ctx, cli, logger, operatorNamespace)
+	if err != nil {
+		logger.Error(err, "failed to get configmap data")
+		return nil, err
+	}
```

Comment:
This need to be provided from the outside as there is no point in reading and parsing the same source map in multiple places